flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -598,12 +599,10 @@ private void registerJobMetrics() {
         final FlinkException cause = new FlinkException("Scheduler is being stopped.");
 
         final CompletableFuture<Void> checkpointServicesShutdownFuture =
-                CompletableFuture.allOf(
-                        executionGraph
-                                .getTerminationFuture()
-                                .thenAcceptAsync(
-                                        this::shutDownCheckpointServices, getMainThreadExecutor()),
-                        checkpointsCleaner.closeAsync());
+                executionGraph
+                        .getTerminationFuture()
+                        .thenAcceptAsync(this::shutDownCheckpointServices, getMainThreadExecutor())
+                        .thenCompose(ignored -> checkpointsCleaner.closeAsync());
```

Comment:
I would probably use `FutureUtils.composeAfterwards` because this will run `checkpointsCleaner.closeAsync()` also in case that the previous future fails and forward the exception afterwards.

Reply:
We should then do the same for shutdown of checkpoint services (the same way as we do in adaptive scheduler).

Reply:
👍